python_modules/libraries/dagster-dbt/dagster_dbt/cloud/asset_defs.py

```diff
@@ -213,8 +200,55 @@ def _get_dbt_nodes_and_dependencies(
         if self._has_generate_docs:
             compile_job_materialization_command_step -= 1
 
-        # Fetch the compilation run's manifest and run results.
-        compile_run_id = compile_run_dbt_output.run_id
+        return compile_run_dbt_output.run_id, compile_job_materialization_command_step
+
+    def _get_dbt_nodes_and_dependencies(
+        self,
+    ) -> Tuple[Mapping[str, Any], Mapping[str, FrozenSet[str]]]:
+        """
+        For a given dbt Cloud job, fetch the latest run's dependency structure of executed nodes.
+        """
+        # Fetch information about the job.
+        job = self._dbt_cloud.get_job(job_id=self._job_id)
+        self._project_id = job["project_id"]
+        self._has_generate_docs = job["generate_docs"]
+
+        # We constraint the kinds of dbt Cloud jobs that we support running.
+        #
+        # A simple constraint is that we only support jobs that run multiple steps,
+        # but it must contain one of either `dbt run` or `dbt build`.
+        #
+        # As a reminder, `dbt deps` is automatically run before the job's configured commands.
+        # And if the settings are enabled, `dbt docs generate` and `dbt source freshness` can
+        # automatically run after the job's configured commands.
+        #
+        # These commands that execute before and after the job's configured commands do not count
+        # towards the single command constraint.
+        self._job_commands = job["execute_steps"]
+        self._job_materialization_command_step = (
+            DbtCloudCacheableAssetsDefinition.get_job_materialization_command_step(
+                execute_steps=self._job_commands
+            )
+        )
+
+        # Determine whether to use a cached compile run. This should only be set up if the user is
+        # using a GitHub action along with their dbt project.
+        dbt_cloud_job_env_vars = self._dbt_cloud.get_job_environment_variables(
+            project_id=self._project_id, job_id=self._job_id
+        )
+        compile_run_id = (
+            dbt_cloud_job_env_vars.get(DAGSTER_DBT_COMPILE_RUN_ID_ENV_VAR, {})
+            .get("job", {})
+            .get("value")
+        )
+
+        compile_run_id, compile_job_materialization_command_step = (
+            # If a compile run is cached, then use it.
+            self._get_cached_compile_dbt_cloud_job_run(compile_run_id=int(compile_run_id))
```

Comment:
not blocking, and mostly a matter of taste, but I think this might be easier to understand if there was a single self._get_compile_run_id_and_command_step (or self._get_manifest_and_run_results) function which branched its behavior internally

Reply:
as an aside, I wasn't aware that a compile command would generate a run_results.json file

Reply:
We're going through dark magic here in the dbt Cloud APIs 🥲